### ISSUE TITLE
apptrack handling

### ISSUE THREAD

**Issue by caver456** [ncssar/sartopo_bg#17]
According to the docs, as soon as a mobile user finishes their recording, the apptrack will be changed to a line on the map they are recording to.

So, apptrack features are either recordings still in progress (whether during an assignment, or long afterwards if the searcher forgot to finish the recording) or were imported from a json that was exported while recording was in progress.

One could say that apptracks on a map should be exported to json as lines, and/or apptracks in json should be imported to the map as lines.

Either way: how should dmg interact with apptracks?

Initial ideas:

There definitely are still instances of the same data being both an apptrack and a line, so, we need to check for that case too.

[side note: 'truncate here' is probably not useful here - it will actually remove the >earlier< portion of an apptrack or locator tail, and preserve everything more recent than the clicked location.  See help ticket #1368, 1/4/2022) The wording is dodgy if not completely opposite from the usual interpretation: 'truncate' usually means that the point you are clicking should become the >end< of the apptrack and everything more recent should be discarded, but apparently it means that the point you are clicking should become the >beginning< of the apptrack, and everything more recent will be preserved.  If you click 'truncate here' at the expected end of the searcher's apptrack, i.e. when they left the CP to go home at the end of the day after their search, you will get a nasty surprise, and there is no undo.]

To reduce pointless map requests during the search, maybe the best thing to do is to just keep a list of the apptrack names that are currently in process but don't try to update them on the debrief map on every sync.  Ideally, at the end of the search, all apptracks have been converted to lines, so that list is empty.

The list of apptracks will allow the operator to ask that searcher to finish their track, if they are at CP; but, to allow the operator to clean things up if the searcher is not at CP (maybe they have gone home and forgot to finish their track, or maybe they have finished their search and are enroute to CP), the list of apptracks should also have an option for each apptrack to 'copy now to debrief map as line'.

If this action bypasses sync and copies straight to the debrief map, so that the incident map apptrack is unaffected, then it's possible the apptrack could be finished later at which point it would become a line and picked up by sync.  The dmg operator should then be presented with options on how to handle the duplicate lines (presumably of the same name).  Hopefully, all points from the old line are also points from the new line, in which case the old line can be deleted and the new line can be imported.

**Comment by caver456** [comment]
Another thought, to reduce the need for operator interaction:

Deal with apptracks when the Generate PDF button is clicked for a given outing.  Do this automatically if possible: copy any unique apptracks-in-process (i.e. apptracks whose points are not coincident with any other line) to lines in the debrief map, as above, then generate the PDF as normal.

In the table, we really want to display a few different tracks counts: 1) finished tracks (from finished apptracks or from Garmin or other means like a file someone sent us); 2) unique-apptracks-in-process.  Maybe the best / most succinct and compact way to display this is to just stick with one column named 'Tracks' but have hybrid cell values:

`<number-of-finished-tracks>[ + <number-of-unique-apptracks-in-process>]`

example: 

`1`

means there is one finished track and no unique-apptracks-in-process

`3 + 2`

means there are three finished tracks and two unique-apptracks-in-process

But - how should the code determine what outing a given apptrack should be associated with?  Apptrack title is the only automatic option.

So, how to deal with apptracks that don't have appropriate track titles?

All apptracks without proper titles could be listed in a separate unclaimed apptracks listing / dialog, which allows the operator to manually associate each one with an outing.

Automatic reduction in craziness as the search progresses:

When an unclaimed apptrack-in-process gets renamed (at any time while dmg is running) with an appropriate track name:

1) the unclaimed apptrack listing disappears and the unclaimed apptrack count reduces by one

AND one of these:

2a) if the unclaimed apptrack was already associated manually with an outing, then that outing's track numbers wouldn't change (3+2 would remain as 3+2)

OR

2b) if the unclaimed apptrack had not been manually associated with an outing then the outing's associated-unique-apptracks-in-process count would increase by one.

Eventually, the goal is for the track numbers to not have any unique-apptracks-in-process - all searchers should finish their apptracks at which point they become lines.

(We know that the finished apptracks could also remain as apptracks, which is why the unique-apptracks (without points coincident to any line) checking is needed.  Coincident apptracks should be silently ignored as long as all of their points are coincident with an existing line.)

This also addresses the issue about spurious apptracks showing up on debrief maps, with searcher names and sar numbers etc: if an apptrack hasn't been associated with an outing, either automatically (due to having an appropriate track name) or manually by the operator as described, then, don't include it in generated PDFs.

**Comment by caver456** [comment]
Unambiguous task: write a function to compare an apptrack and a line - return a value that indicates whether a) the line as a subset of the apptrack, b) the apptrack is a subset of the line, or c) neither.

If the line is a subset of the apptrack (all points of the line are also points of the apptrack in the same sequence - though the apptrack could be longer on either end) then the line can safely be ignored; if the apptrack is a subset of the line then the apptrack could safely be ignored.

This should really only be the case for the buggy finished apptracks that remain as apptracks as well as lines, since the docs say that all finished apptracks should automatically be converted to lines.

Example in training.json - an apptrack and a line, both with the same title AND the same id!

```
                "type": "LineString"
            },
            "id": "3964f94c-5192-4e2a-9e1d-8c32debba97a",
            "type": "Feature",
            "properties": {
                "stroke-opacity": 1,
                "creator": "CUCGN5",
                "pattern": "solid",
                "stroke-width": 2,
                "title": "Ac101C",
                "fill": "#ff0000",
                "stroke": "#ff0000",
                "fill-opacity": 0.1,
                "class": "Shape",
                "updated": 1636905034000,
                "gpstype": "TRACK"
            }

...

                "type": "LineString"
            },
            "id": "3964f94c-5192-4e2a-9e1d-8c32debba97a",
            "type": "Feature",
            "properties": {
                "stroke-opacity": 1,
                "creator": "CUCGN5",
                "pattern": "solid",
                "stroke-width": 2,
                "title": "Ac101C",
                "stroke": "#ff0000",
                "class": "AppTrack",
                "updated": 1636777162000
            }
        },
```

On the map - the apptrack is a red thin line; the line is thick blue; you can see the apptrack is a small subset of the line, so, we need the function to indicate as such:

![image](https://user-images.githubusercontent.com/18752102/157895102-ceb62925-4da3-4d67-a238-1a9ec83d360d.png)

**Comment by caver456** [comment]
Tkidiff (winmerge) of the line geometry vs the apptrack geometry shows some very interesting things:

- there are a lot of sequences of completely identical coordinates - between two and ten (or so) reports of the exact same coordinates and elevation.  Who knows if it's an issue with the app or something else in the phone, but, one wonders how much data could be saved by skipping these duplicates, how many of these sequences exist, how frequently they occur, how long they last, etc.  Will write some code to analyze this.
- it's clear that the apptrack is >mostly< a subset, but there are some differences in the number of duplicates between apptrack and line.  Might need to remove dups before checking.

A few things visible in this winmerge clip (line on left, apptrack on right):
- the line has a report at 355060 but the apptrack has no report at that time... how did it get into the line??
- this sequence has duplicate coordinates but with different elevation (1228 followed by 1227):
![image](https://user-images.githubusercontent.com/18752102/157898103-9b30e42d-f786-4d41-b8f6-1bb3fb35aacd.png)

**Comment by caver456** [comment]
Wow:

```
2022-03-11 09:47:25,996 Analyzing file training.json
2022-03-11 09:47:25,996  A = total duplicate count
2022-03-11 09:47:25,997  B = number of runs of duplicates
2022-03-11 09:47:25,998  C = number of duplicate entries in longest run of duplicates
2022-03-11 09:47:25,998 Class        | Title            | A    | B   | C   |
2022-03-11 09:47:25,999 --------------------------------------------------------------
2022-03-11 09:47:26,162 Shape        | AC101G           |   63 |  19 |  11 |
2022-03-11 09:47:26,164 Shape        | Ac101C           |  727 | 137 |  48 |
2022-03-11 09:47:26,166 Shape        | AJ104E           |   10 |   9 |   2 |
2022-03-11 09:47:26,167 Shape        | AB102D           |  475 | 162 |  15 |
2022-03-11 09:47:26,168 Shape        | AA102F           |  195 |  77 |  14 |
2022-03-11 09:47:26,170 Shape        | AA102D           |  138 |  50 |  10 |
2022-03-11 09:47:26,171 Shape        | <name>129        |    1 |   1 |   1 |
2022-03-11 09:47:26,171 Shape        | AI103a           |    1 |   1 |   1 |
2022-03-11 09:47:26,172 Shape        | AB102F           |  308 |  94 |  21 |
2022-03-11 09:47:26,176 Shape        | AC101B           |  401 | 194 |  20 |
2022-03-11 09:47:26,178 Shape        | AB102C           |    1 |   1 |   1 |
2022-03-11 09:47:26,182 Shape        | <name> 57        |  435 |  15 | 116 |
2022-03-11 09:47:26,183 Shape        | AJ104D           |    1 |   1 |   1 |
2022-03-11 09:47:26,185 Shape        | AI103f           |  404 | 157 |  18 |
2022-03-11 09:47:26,186 Shape        | AI               |   60 |  23 |   8 |
2022-03-11 09:47:26,187 Shape        | AA102E           |   74 |  40 |  10 |
2022-03-11 09:47:26,189 Shape        | AI103c           |    1 |   1 |   1 |
2022-03-11 09:47:26,190 Shape        | AI103b           |  226 |  74 |  53 |
2022-03-11 09:47:26,191 Shape        | Ai103d           |   36 |  19 |  10 |
2022-03-11 09:47:26,193 Shape        | Ac101D           |  480 | 105 |  30 |
2022-03-11 09:47:26,194 Shape        | AJ-104-F         |   10 |   5 |   4 |
2022-03-11 09:47:26,198 AppTrack     | AB102            | 1928 | 230 | 517 |
2022-03-11 09:47:26,199 AppTrack     | Ac101C           |  257 |  55 |  44 |
2022-03-11 09:47:26,201 AppTrack     | AC101a           |   62 |  28 |   6 |
2022-03-11 09:47:26,201 AppTrack     | AC101E           |  351 |  63 |  76 |
2022-03-11 09:47:26,203 AppTrack     | <name>72         |  223 |  96 |   8 |
2022-03-11 09:47:26,204 AppTrack     | AI               |   57 |  22 |   8 |
2022-03-11 09:47:26,205 AppTrack     | AA102G           |  987 |  95 | 474 |
2022-03-11 09:47:26,205 AppTrack     | AD105B           |  167 |  50 |  72 |
2022-03-11 09:47:26,207 AppTrack     | AI103b           |  239 |  74 |  53 |
2022-03-11 09:47:26,209 AppTrack     | Ai103d           |   36 |  20 |  10 |
2022-03-11 09:47:26,211 AppTrack     | Ac101D           |  523 | 108 |  37 |
2022-03-11 09:47:26,212 AppTrack     | AA102A           |    5 |   5 |   1 |
2022-03-11 09:47:26,212 Locator      | SAR 66           |    1 |   1 |   1 |
2022-03-11 09:47:26,213 Locator      | SAR 81           |    1 |   1 |   1 |
2022-03-11 09:47:26,213 Locator      | SAR 89           |    4 |   2 |   2 |
2022-03-11 09:47:26,215 Locator      | SAR 98           |    1 |   1 |   1 |
2022-03-11 09:47:26,215 Locator      | SAR 106          |    1 |   1 |   1 |
2022-03-11 09:47:26,216 Locator      | Team 101         |   13 |   7 |   5 |
2022-03-11 09:47:26,217 Locator      | Team 102         |    3 |   2 |   2 |
2022-03-11 09:47:26,217 Locator      | Team 103         |    1 |   1 |   1 |
2022-03-11 09:47:26,218 Locator      | Team 104         |    2 |   1 |   2 |
2022-03-11 09:47:26,218 Locator      | SAR 26           |    1 |   1 |   1 |
2022-03-11 09:47:26,219 Locator      | Team             |    1 |   1 |   1 |
2022-03-11 09:47:26,220 --------------------------------------------------------------
2022-03-11 09:47:26,220  Total duplicates in this file: 8911
2022-03-11 09:47:26,220  Total estimated file size used by duplicates (60 bytes each): 522.1kB
2022-03-11 09:47:26,221    (Actual file size = 4795.6kB --> 10.9% used by duplicates)
2022-03-11 09:47:26,221 End of summary.
```

**Comment by caver456** [comment]
Tried this code to remove runs of repeated points:

dsgc=[k for k,g in itertools.groupby([p[0:2] for p in dupShape['geometry']['coordinates']])]
dagc=[k for k,g in itertools.groupby([p[0:2] for p in dupApptrack['geometry']['coordinates']])]

which appears to work, but, afterwards, neither list is a sublist of the other.  Dumping each to a file then tkdiffing shows that the line has several points where the apptrack has holes in its sequence of timestamps (the yellow point in the shape is from timestamp ...187072 which is missing from the apptrack):

![image](https://user-images.githubusercontent.com/18752102/157945280-bf3963b2-8db1-451f-a735-c5d32de355f7.png)

This is one of maybe 50 or so such glitches.  In all cases, the line has points where the apptrack has gaps.  The longest run of these is 5 points; most are just one point.

Bottom line: we still want to be able to say that the apptrack is a sublist of the line.  Maybe if we check to make sure that each differing point in the line corresponds with a hole in the apptrack timeline, by only comparing points that have matching timestamps?

**Comment by caver456** [comment]
Got it working as described above: only compare the coordinates for which a timestamp exists in both coordinate lists.  Converting them to dictionaries, where keys are timestamps, makes it much easier since there's no need to compare sequence:

```
def isApptrackSubsetOfLine(lineCoords,apptrackCoords):
    # comparisons will be easier if we make a dictionary: keys are timestamps, values are [lon,lat] (ignore elevation)
    #  but we need to do a sort again before it's readable
    lineTSDict={v[3]:v[0:2] for v in lineCoords}
    apptrackTSDict={v[3]:v[0:2] for v in apptrackCoords}
    # this could be a one-liner since we only need to check if one dict is subset of the other (sequence doesn't matter)
    #  but this way we have access to internal statistics if needed
    subset=True
    matches=0
    tsDifferencesList=[]
    lineTsMissingFromApptrackList=[]
    for ts in lineTSDict.keys():
        if ts in apptrackTSDict.keys():
            if lineTSDict[ts]==apptrackTSDict[ts]:
                matches+=1
            else:
                subset=False
                tsDifferencesList.append(ts)
        else:
            lineTsMissingFromApptrackList.append(ts)
    if matches==0:
        subset=False
    # logging.info('matches:'+str(matches))
    # logging.info('tsDifferences:'+str(len(tsDifferencesList)))
    # # logging.info('lineTsMissingFromApptrackList:'+str(len(lineTsMissingFromApptrackList))+':'+str(lineTsMissingFromApptrackList))
    # logging.info('lineTsMissingFromApptrackList:'+str(len(lineTsMissingFromApptrackList)))
    return subset
```

Interesting side note: in training.json, same-ID correlated perfectly with is-subset:

```
2022-03-12 06:32:33,539 Shape has same name as AppTrack:Ac101C
2022-03-12 06:32:33,539   Do they have the same id? YES (3964... 3964...)
2022-03-12 06:32:33,541   is apptrack subset of line:True
2022-03-12 06:32:33,541 Shape has same name as AppTrack:AD105C
2022-03-12 06:32:33,541   Do they have the same id? NO (8cc2... 06d8...)
2022-03-12 06:32:33,542   is apptrack subset of line:False
2022-03-12 06:32:33,542 Shape has same name as AppTrack:AI
2022-03-12 06:32:33,542   Do they have the same id? YES (aeaf... aeaf...)
2022-03-12 06:32:33,543   is apptrack subset of line:True
2022-03-12 06:32:33,544 Shape has same name as AppTrack:AI103b
2022-03-12 06:32:33,544   Do they have the same id? YES (c466... c466...)
2022-03-12 06:32:33,544   is apptrack subset of line:True
2022-03-12 06:32:33,545 Shape has same name as AppTrack:Ai103d
2022-03-12 06:32:33,545   Do they have the same id? YES (d36a... d36a...)
2022-03-12 06:32:33,548   is apptrack subset of line:True
2022-03-12 06:32:33,548 Shape has same name as AppTrack:Ac101D
2022-03-12 06:32:33,549   Do they have the same id? YES (d55b... d55b...)
2022-03-12 06:32:33,550   is apptrack subset of line:True
```

**Comment by caver456** [comment]
Got a good start with this GUI:
![image](https://user-images.githubusercontent.com/18752102/158073549-c186fb51-7e4a-4339-92ec-278e29ed378c.png)

But, it also occurs that we want to omit lines that used to be apptracks and aren't associated with an outing, like the finished apptracks of searchers enroute to the search, like Smith123.  That will be a different - though closely related - issue.

**Comment by caver456** [comment]
After emails, decided to back off and make this a read-only dialog, mainly due to simplicity - changing the outing associations from this dialog means the apptrack name would have to be / should be changed as well; the operator would have to choose a suffix letter; many interactions to check; so, the alternative is to just require the operator to change the apptrack name in the map in order to affect its association with an outing.

Anyway - good progress, but, the apptrack rename process isn't getting handled correctly.

Looking at the sequence of sync responses when an apptrack name is changed from the phone:

- in each response, geometry only has one four-element coordinate, and also has the correct size value.
- properties exists on each response even if there geometry was identical to the previous response (i.e. there was no coordinate update) - maybe just to indicate that the apptrack is still being recorded in which case it has 'nop':true
- following rename, properties will contain the full set of properties including 'title'; so, if properties.title exists, change dmd to match, and check the new title to see if it should be associated with an outing.

**Comment by caver456** [comment]
Got the name change processing to work correctly as above.

On to the question of getting apptracks-in-process to appear on generated PDFs.

Options:
1) add an uncropped line to the debrief map whose coordinates are those of the app track so far (then crop it)
2) don't add any line to the debrief map; just add the json directly to the pdf request.


1 has the complications of deleting the old line on subsequent pdf gen, and/or when the apptrack is finished and converted to a line.  How to accomplish this idea of a 'temporary track'?
2 avoids those complications.  The apptrack never becomes a feature on the actual sartopo debrief map, until it's finished (as a line). 

2 sure sounds nicer.  The only disadvantage is that the sartopo debrief map is not 'up to date', in the sense that it won't show apptracks until they are finished on the incident map.  That is significant if anyone actually views the debrief map, but, if they do, it should be obvious due to so much missing expected data.

Also, with 1, the question comes up of when you would copy the unfinished apptrack from the incident map to the debrief map.  Unless you update the debrief map with every apptrack update - which is a lot of requests for purely redundant data (redundant with the incident map), then you have to pick discrete time(s) to do the 'import'.  The most obvious is whenever the Generate PDF button is clicked.  But, that could be misleading: 'apptracks' on the debrief map could potentially be out of date after the PDF is generated.

Overall it's probably best to go with 2, and to just stick to the policy that the debrief map itself won't show searcher tracks until those apptracks have been finished.

2 means we need to keep the cache up to date when apptrack updates are in the response - we just don't need to turn them into debrief map requests.

**Comment by caver456** [comment]
Went with 2: don't add a debrief map feature, just add the json directly to the print request, with a heavy-dashed line style.
